worldengine/draw.py

```diff
@@ -209,26 +393,93 @@ def draw_riversmap(world, target):
+    #  and additional "business logic" defined in get_biome_color_based_on_elevation
     for y in range(world.height):
         for x in range(world.width):
-            target.set_pixel(x, y, (c[y, x], c[y, x], c[y, x], 255))
+            # Get the normalized elevation at this pixel
+            elev = elevation_mask[y, x]
+            
+            # Get a rgb noise value, with some logic to modify it based on the elevation of the tile
+            r, g, b = get_biome_color_based_on_elevation(world, elev, x, y)
+
+            # Set pixel to this color. This initial color will be accessed and modified later when 
+            # the map is smoothed and shaded.
+            target.set_pixel(x, y, (r, g, b, 255))
+
+
+    # Loop through and average a pixel with its neighbors to smooth transitions between biomes
+    for y in range(1, world.height-1):
+        for x in range(1, world.width-1):
+            ## Only smooth land tiles
+            if world.is_land((x, y)):
+                # Lists to hold the separated rgb values of the neighboring pixels
+                all_r = []
+                all_g = []
+                all_b = []
+
+                # Loop through this pixel and all neighboring pixels
+                for j in range(y-1, y+2):
+                    for i in range(x-1, x+2):
+                        # Don't include ocean in the smoothing, if this tile happens to border an ocean
+                        if world.is_land((i, j)):
+                            # Grab each rgb value and append to the list
+                            r, g, b, a = target.pixels[i, j]
+                            all_r.append(r)
+                            all_g.append(g)
+                            all_b.append(b)
+
+                # Making sure there is at least one valid tile to be smoothed before we attempt to average the values
+                if all_r:
+                    avg_r = int(sum(all_r) / len(all_r))
+                    avg_g = int(sum(all_g) / len(all_g))
+                    avg_b = int(sum(all_b) / len(all_b))
+
+                    ## Setting color of the pixel again - this will be once more modified by the shading algorithm
+                    target.set_pixel(x, y, (avg_r, avg_g, avg_b, 255))
+
+    # "Shade" the map by sending beams of light west to east, and increasing or decreasing value of pixel based on elevation difference
+    for y in range(SAT_SHADOW_SIZE-1, world.height-SAT_SHADOW_SIZE-1):
+        for x in range(SAT_SHADOW_SIZE-1, world.width-SAT_SHADOW_SIZE-1):
+            if world.is_land((x, y)):
+                r, g, b, a = target.pixels[x, y]
+                
+                # Build up list of elevations in the previous n tiles, where n is the shadow size.
+                # This goes left to right, so it would be the previous tiles on the same y level  
+                prev_elevs = [ world.elevation['data'][y-n, x-n] for n in range(1, SAT_SHADOW_SIZE+1) ]
```

Comment:
You should probably use slice notation here, it's faster and will return a numpy array, should you need it.